lightning/src/ln/msgs.rs

```diff
@@ -52,49 +53,121 @@ pub enum DecodeError {
 	Io(::std::io::Error),
 }
 
-/// Tracks localfeatures which are only in init messages
-#[derive(Clone, PartialEq)]
-pub struct LocalFeatures {
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in the init message, sent between peers and not
+/// rumored around the P2P network.
+pub struct FeatureContextInit {}
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in the node_announcement message, as it is
+/// rumored around the P2P network.
+pub struct FeatureContextNode {}
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in the ChannelAnnouncement message, as it is
+/// rumored around the P2P network.
+pub struct FeatureContextChannel {}
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in an invoice, used to determine the different
+/// options available for routing a payment.
+///
+/// Note that this is currently unused as invoices come to us via a different crate and are not
+/// native to rust-lightning directly.
+pub struct FeatureContextInvoice {}
+
+/// An internal trait capturing the various future context types
+pub trait FeatureContext {}
+impl FeatureContext for FeatureContextInit {}
+impl FeatureContext for FeatureContextNode {}
+impl FeatureContext for FeatureContextChannel {}
+impl FeatureContext for FeatureContextInvoice {}
+
+/// An internal trait capturing FeatureContextInit and FeatureContextNode
+pub trait FeatureContextInitNode : FeatureContext {}
+impl FeatureContextInitNode for FeatureContextInit {}
+impl FeatureContextInitNode for FeatureContextNode {}
```

Comment:
BOLT 9 uses groups of contexts to specify when features are presented. However, this combination of contexts (e.g., IN) seems more like a convenience used in the tabular specification of features. Rather than reflecting these combinations directly as traits, the feature itself could be a trait in which each context provides an (empty) implementation. The code organization could reflect the table:

```rust
pub struct InitContext {}
pub struct NodeAnnouncementContext {}
pub struct ChannelAnnouncementContext {}

pub trait OptionDataLossProtect {}
impl OptionDataLossProtect for InitContext {}
impl OptionDataLossProtect for NodeAnnouncementContext {}

pub trait InitialRoutingSync {}
impl InitialRoutingSync for InitContext {}

pub trait OptionUpfrontShutdownScript {}
impl OptionUpfrontShutdownScript for InitContext {}
impl OptionUpfrontShutdownScript for NodeAnnouncementContext {}
```

Methods would be defined on `Features` using trait bounds similarly to how you are already doing only where the bound is the "feature" trait rather than "contexts" trait.